lib/ddtrace/trace_operation.rb

```diff
@@ -0,0 +1,400 @@
+      resource: nil,
+      rule_sample_rate: nil,
+      runtime_id: nil,
+      sample_rate: nil,
+      sampled: nil,
+      sampling_priority: nil,
+      service: nil
+    )
+      # Attributes
+      @events = events || Events.new
+      @id = id || Utils.next_id
+      @max_length = max_length || DEFAULT_MAX_LENGTH
+      @parent_span_id = parent_span_id || 0
+      @sampled = sampled.nil? ? false : sampled
+
+      # Tags
+      @agent_sample_rate = agent_sample_rate
+
+      # TODO: Move hostname resolution elsewhere?
+      #       Don't want trace to know about configuration.
+      hostname = Core::Environment::Socket.hostname if Datadog.configuration.report_hostname
+      @hostname = hostname && hostname.empty? ? nil : hostname
+
+      @lang = Core::Environment::Identity.lang
+      @name = name
+      @origin = origin
+      @process_id = Datadog::Core::Environment::Identity.pid
+      @rate_limiter_rate = rate_limiter_rate
+      @resource = resource
+      @rule_sample_rate = rule_sample_rate
+      @runtime_id = Datadog::Core::Environment::Identity.id
+      @sample_rate = sample_rate
+      @sampling_priority = sampling_priority
+      @service = service
+
+      # State
+      @root_span = nil
+      @active_span = nil
+      @active_span_count = 0
+      @events = events || Events.new
+      @finished = false
+      @spans = []
+    end
+
+    # TODO: Maybe not the right strategy for preventing
+    #       big traces? Prevents new spans from starting if
+    #       too many spans are finished. Has no effect if all
+    #       spans are in-flight... likely needs to change.
+    def full?
+      @max_length > 0 && @active_span_count >= @max_length
+    end
+
+    def finished_span_count
+      @spans.length
+    end
+
+    def finished?
+      @finished == true
+    end
+
+    def sampled?
+      @sampled == true || (!@sampling_priority.nil? && @sampling_priority > 0)
+    end
+
+    def keep!
+      self.sampling_priority = Datadog::Ext::Priority::USER_KEEP
+    end
+
+    def reject!
+      self.sampling_priority = Datadog::Ext::Priority::USER_REJECT
+    end
+
+    def measure(op_name, **span_options, &block)
+      # Don't allow more span measurements if the
+      # trace is already completed. Prevents multiple
+      # root spans with parent_span_id = 0.
+      return yield(SpanOperation.new(op_name), TraceOperation.new) if finished? || full?
+
+      # Create new span
+      span_op = build_span(op_name, **span_options)
+
+      # Start span measurement
+      span_op.measure { |s| yield(s, self) }
+    end
+
+    def build_span(op_name, **span_options)
+      begin
+        # Resolve span options:
+        # Context, parent, service name, etc.
+        span_options = build_span_options(**span_options)
+
+        # Build a new span operation
+        Datadog::SpanOperation.new(op_name, **span_options)
+      rescue StandardError => e
+        Datadog.logger.debug("Failed to build new span: #{e}")
```

Comment:
Minor: Perhaps convert this to a block, to avoid rendering the whole string when debug logging isn't active?

(Same applies to other places in this class, and possibly in this PR)

Reply:
Yeah, I can take a look at that.